src/beast/beast/http/empty_body.h

```diff
@@ -0,0 +1,103 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of Beast: https://github.com/vinniefalco/Beast
+    Copyright 2013, Vinnie Falco <vinnie.falco@gmail.com>
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef BEAST_HTTP_EMPTY_BODY_H_INCLUDED
+#define BEAST_HTTP_EMPTY_BODY_H_INCLUDED
+
+#include <beast/http/error.h>
+#include <beast/http/message.h>
+#include <beast/asio/streambuf.h>
+#include <memory>
+#include <string>
+
+namespace beast {
+namespace http {
+
+/** An empty content-body.
+*/
+struct empty_body
+{
+    struct value_type
+    {
+    };
+
+    struct reader
+    {
+        template<bool isReq, class Allocator>
+        explicit
+        reader(message<isReq, empty_body, Allocator>&)
+        {
+        }
+
+        void
+        write(void const*, std::size_t, error_code&)
+        {
+        }
+    };
+
+    class writer
+    {
+        streambuf sb;
+
+    public:
+        static bool constexpr is_single_pass = true;
+    
+        template<bool isReq, class Allocator>
+        explicit
+        writer(message<isReq, empty_body, Allocator> const& m)
+        {
+            m.write_headers(sb);
+        }
+
+        void
+        init(error_code&)
+        {
+        }
+
+        auto
+        data() const
+        {
+            return sb.data();
+        }
+    };
+
+    template<bool isRequest, class Allocator>
+    static
+    void
+    prepare(prepared_message<isRequest, empty_body, Allocator>& msg)
+    {
+        msg.headers.replace("Content-Length", 0);
+        msg.headers.erase("Content-Type"); // VFALCO this right?
```

Comment:
I think this is right. https://tools.ietf.org/html/rfc7231#section-3.1.1.5 says

> A sender that generates a message containing a payload body SHOULD
   generate a Content-Type header field in that message unless the
   intended media type of the enclosed representation is unknown to the
   sender. 

Since there's no message body, the "SHOULD" doesn't apply. ("SHOULD" is also not as strong as "SHALL" or "MUST".)

> If a Content-Type header field is not present, the recipient
   MAY either assume a media type of "application/octet-stream"
   ([RFC2046], Section 4.5.1) or examine the data to determine its type.

And here's what the recipient is supposed to do if it's not present...